finality-aleph/src/finalization.rs

```diff
@@ -77,194 +80,133 @@ where
-    if new_data.hash == last_finalized {
-        return None;
+    let first_block_in_session = session_boundaries.first_block;
+    let last_block_in_session = session_boundaries.last_block;
+    if proposal.branch.is_empty() {
+        error!(target:"afa", "Incorrect proposal {:?} passed through data availability", proposal);
+        return false;
     }
-
-    if new_data.number > last_block_in_session {
-        return None;
+    if proposal.branch.len() > MAX_DATA_BRANCH_LEN as usize {
+        error!(target:"afa", "Incorrect proposal {:?} passed through data availability", proposal);
+        return false;
     }
 
-    let last_finalized_number = match client.number(last_finalized) {
-        Ok(Some(number)) => number,
-        _ => {
-            error!(target: "afa", "No block number for {}", last_finalized);
-            return None;
-        }
-    };
-
-    if let Ok(Some(header)) = client.header(BlockId::Hash(new_data.hash)) {
-        if *header.number() != new_data.number {
-            warn!(target: "afa", "Incorrect number for hash {}. Got {}, should be {}", new_data.hash, new_data.number, header.number());
-            return None;
-        }
-    } else {
-        warn!(target: "afa", "No header for hash {}", new_data.hash);
-        return None;
+    // Since `proposal.branch.len() == 0` is ruled out above, so we can safely do `proposal.branch.len() - 1`
+    let branch_base_num =
+        proposal.number - <NumberFor<B>>::saturated_from(proposal.branch.len() - 1);
+    if branch_base_num < first_block_in_session {
+        warn!(target:"afa", "Incorrect proposal {:?}. Too low num {:?} first_block {:?}", proposal, proposal.number, first_block_in_session);
+        return false;
+    }
+    if proposal.number > last_block_in_session {
+        warn!(target:"afa", "Incorrect proposal {:?}. Too high num {:?} last_block {:?}", proposal, proposal.number, last_block_in_session);
+        return false;
     }
 
-    // Iterate ancestors of `new_hash` until reaching a block with number <= last_finalized_number
-    // in order to check if new_data.hash is an ancestor of last_finalized
-    let mut hash = new_data.hash;
-    loop {
-        let header = match client.header(BlockId::Hash(hash)) {
-            Ok(Some(header)) => header,
-            _ => {
-                error!(target: "afa", "No header for hash {}", hash);
-                return None;
-            }
-        };
-
-        if header.number() <= &last_finalized_number {
-            if hash != last_finalized {
-                // `new_hash` is not an ancestor of `last_finalized`
-                return None;
+    // For each block in the branch we check that:
+    // 1) It is available
+    // 2) It's parent is its predecessor on the branch
+    // 3) The number of the first (bottom) block in the branch is as claimed in `proposal`.
+    // If any of the checks fails, we return `false`.
+
+    for i in 0..proposal.branch.len() {
```

Comment:
Before the for:
```
match client
            .header(BlockId::Hash(proposal.branch[0]))
            .expect("client should not fail") {
            None => {
                    warn!(target: "afa", "Block {:?} in proposal not imported.", curr_hash);
                    return false;
            },
            Some(header) => if *header.number() != branch_base_num {
                    warn!(target: "afa", "Incorrect num {:?} claimed for {:?}. Correct is {:?}.", proposal.number, curr_hash, header.number());
                    return false;
             }
}
```
The for itself:
```
for (block, parent) in proposal.branch.iter().skip(1).zip(proposal.branch) {
          // only the parent validation logic
}
```

Bleh, after writing this I noticed that if you rewrite this all into methods of proposals then this will be much less useful... oh well I hope it'll be at least a bit.
Considering that this is the second instance of this maybe it

Reply:
Changed. Take a look if it is correct now, as I had to make some changes for it to compile.